dbms/src/Interpreters/InterpreterInsertQuery.cpp

```diff
@@ -99,73 +99,99 @@ BlockIO InterpreterInsertQuery::execute()
     const auto & query = query_ptr->as<ASTInsertQuery &>();
     checkAccess(query);
 
+    BlockIO res;
     StoragePtr table = getTable(query);
 
     auto table_lock = table->lockStructureForShare(true, context.getInitialQueryId());
 
-    /// We create a pipeline of several streams, into which we will write data.
-    BlockOutputStreamPtr out;
-
-    /// NOTE: we explicitly ignore bound materialized views when inserting into Kafka Storage.
-    ///       Otherwise we'll get duplicates when MV reads same rows again from Kafka.
-    if (table->noPushingToViews() && !no_destination)
-        out = table->write(query_ptr, context);
-    else
-        out = std::make_shared<PushingToViewsBlockOutputStream>(query.database, query.table, table, context, query_ptr, no_destination);
-
-    /// Do not squash blocks if it is a sync INSERT into Distributed, since it lead to double bufferization on client and server side.
-    /// Client-side bufferization might cause excessive timeouts (especially in case of big blocks).
-    if (!(context.getSettingsRef().insert_distributed_sync && table->isRemote()) && !no_squash)
+    BlockInputStreams in_streams;
+    size_t out_streams_size = 1;
+    if (query.select)
     {
-        out = std::make_shared<SquashingBlockOutputStream>(
-            out, out->getHeader(), context.getSettingsRef().min_insert_block_size_rows, context.getSettingsRef().min_insert_block_size_bytes);
+        /// Passing 1 as subquery_depth will disable limiting size of intermediate result.
+        InterpreterSelectWithUnionQuery interpreter_select{query.select, context, SelectQueryOptions(QueryProcessingStage::Complete, 1)};
+
+        if (table->supportsParallelInsert())
+        {
+            in_streams = interpreter_select.executeWithMultipleStreams(res.pipeline);
+            const Settings & settings = context.getSettingsRef();
+            out_streams_size = std::min(size_t(settings.max_insert_threads), in_streams.size());
```

Comment:
I think we need set at least 1 even if `settings.max_insert_threads` is zero.

Reply:
Thanks, it has been implemented in the new commit.